### PR TITLE
Defer work we do with code actions until we actually need to do it.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Note: features that own their own diagnostic IDs don't need to validate the state of the world when they are asked for fixes.  i.e. they should have only produced a diagnostic in the first place if they were going to be able to fix it.

As such, having these features go and check if they're able to find the very consturct they just analyzed is unnecessary and wasteful.  

--

Note: this doesn't apply to features that have to operate on diagnostics IDs they don't own (i.e. compiler errors).  There we have to validate because there is no contract ensuring that the data will be what we expect it to be.

**Comment by CyrusNajmabadi** [comment]
Ping.  Anyone want to take a look?

File: src/Features/CSharp/Portable/InvokeDelegateWithConditionalAccess/InvokeDelegateWithConditionalAccessCodeFixProvider.cs [hunk 1]
```diff
@@ -29,20 +29,17 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
             context.RegisterCodeFix(new MyCodeAction(
                 CSharpFeaturesResources.DelegateInvocationCanBeSimplified,
-                async c => await UpdateDocumentAsync(context).ConfigureAwait(false),
+                c => UpdateDocumentAsync(context.Document, context.Diagnostics.First(), c),
```
Comment (DustinCampbell) [inline src/Features/CSharp/Portable/InvokeDelegateWithConditionalAccess/InvokeDelegateWithConditionalAccessCodeFixProvider.cs]: Good catch!

**Comment by DustinCampbell** [comment]
:+1:

**Pull request merged**